rescript/get_unite_data.py

```diff
@@ -0,0 +1,126 @@
+        '8.2': {'fungi':      {False: '10.15156/BIO/786385', True: '10.15156/BIO/786387'},
+                'eukaryotes': {False: '10.15156/BIO/786386', True: '10.15156/BIO/786388'}},
+        '8.0': {'fungi':      {False: '', True: '10.15156/BIO/786349'},
+                'eukaryotes': {False: '', True: ''}},  # All other 8.0 are in zip files
+    }
+    # There's got to be a better way! See https://stackoverflow.com/questions/25833613/safe-method-to-get-value-of-nested-dictionary
+    try:
+        # Check if we have the DOI requested
+        doi = unite_dois[version][taxon_group][singletons]
+    except KeyError as ke:
+        print('Unknown DOI for this value: ' + str(ke))
+        raise
+    return doi
+
+# Testing
+_unite_get_doi(version='9.0', taxon_group='fungi', singletons=False)
+
+def _unite_get_url(doi):
+    '''Query plutof API for UNITE url'''
+    print('Get URLs for these DOIs:', doi)
+    base_url = 'https://api.plutof.ut.ee/v1/public/dois/'\
+               '?format=vnd.api%2Bjson&identifier='
+    query_data = requests.get(base_url + doi).json()
+    # Updates can be made to files in a DOI, so on the advice of the devs,
+    # only return the last (newest) file with this -1  vv
+    URL = query_data['data'][0]['attributes']['media'][-1]['url']
+    return URL
+
+# Testing
+_unite_get_url(doi = '10.15156/BIO/2938079')
+
+# Make tmp_dir for standalone testing
+# tmp_dir = tempfile.mkdtemp()
+def _unite_get_raw_files(url, download_path):
+    '''Download and extract all fasta and txt files'''
+    response = requests.get(url, stream=True)
+    if response.status_code != 200:
+        raise ValueError("Failed to download the file from " + url)
+    # Save .tgz file
+    unite_file_path = os.path.join(download_path, 'unitefile.tar.gz')
+    with open(unite_file_path, 'wb') as f:
+        f.write(response.content)
+    # Extract only the 'developer' subdirectory
+    with tarfile.open(unite_file_path, 'r:gz') as tar:
+        # Ensure that 'developer' exists in the tar file
+        members = [member for member in tar.getmembers() if member.name.startswith('developer')]
+        if not members:
+            raise ValueError("No 'developer' subdirectory found in the .tar.gz file.")
+        for member in members:
+            # Strip the 'developer' prefix
+            member.name = os.path.basename(member.name)
+            tar.extract(member, path=download_path)
+    return
+
+# Test it by downloading this file
+_unite_get_raw_files(_unite_get_url(doi = '10.15156/BIO/2938079'), tmp_dir)
+
+def _unite_get_qza(cluster_id, download_path):
+    '''
+    Find and import raw files with matching cluster_id
+    
+    Returns: Tuple containing tax_results and seq_results
+    '''
+    tax_results = []
+    seq_results = []
+    # Find all files...
+    for root, dirs, files in os.walk(download_path):
+        # ... with the matching cluster_id
+        filtered_files = [file for file in files if cluster_id in file]
```

Comment:
This works because `cluster_id` is only `99`, `97`, or `dynamic`. 

If cluster_id could collide with part of the data (1/1/1999) this would cause issues

Reply:
I think you need to be more specific. Based on this:

>sh_refs_qiime_ver9_dynamic_25.07.2023.fasta
sh_taxonomy_qiime_ver9_dynamic_25.07.2023.txt

It seems that the file naming schema is something like this:

```
sh_[ refs|taxonomy ] _qiime + _ | ver[version number] + _ + [cluster ID] + _  + date + [.fasta | .txt]
```

So you can try:

```
fnames = ['sh_refs_qiime_ver9_97_25.07.2023.fasta', 
                  'sh_refs_qiime_ver9_99_25.07.2023.fasta', 
                  'sh_refs_qiime_ver9_dynamic_25.07.2023.fasta']

cluster_id = 'dynamic'

for f in fnames:
     clid = f.split('_')[4]
     if clid == cluster_id:
         print('Found what we\'re looking for: ', clid, str(f))
     else:
         print ('Nope, this is not what we want: ', clid, str(f))
```
>Nope, this is not what we want:  97 sh_refs_qiime_ver9_97_25.07.2023.fasta
Nope, this is not what we want:  99 sh_refs_qiime_ver9_99_25.07.2023.fasta
Found what we're looking for:  dynamic sh_refs_qiime_ver9_dynamic_25.07.2023.fasta

Reply:
I've change the filter. Is this what you meant? 
```python
# old
filtered_files = [file for file in files if cluster_id in file]
# New:
filtered_files = [file for file in files if file.split('_')[4]==cluster_id]
```

Thank you for reproducible example, btw

<details><summary>The naming schema is so chaotic</summary>
<p>

It's just like you described, except two optional substrings are inserted after cluster_id.
These show the inclusion of Singletons `s_` and Euks `all_

```txt

sh_[ refs|taxonomy ]_qiime_ver[major_version]_[cluster_id]_[ '' | 's_' ][ '' | 'all_' ] + date + .[fasta|txt]
```

This means `file.split('_')[5]` could be three different things.

I'm not a fan or 'all or nothing' much less  `all_` or ``

</p>
</details>

Reply:
In that case you can use the same approach. First check for the singleton files, if requested, then follow up with a the appropriate slice to get the cluster id. Perhaps make a fun complex nested list comprehension. :-)